appinventor/appengine/src/com/google/appinventor/client/explorer/youngandroid/ProjectList.java

```diff
@@ -319,139 +198,182 @@ public void refreshTable(boolean needToSort, boolean isInTrash) {
+    return folder.getSelectedFolders();
   }
 
+  @UiHandler("selectAllCheckBox")
+  void toggleAllItemSelection(ClickEvent e) {
+    folder.selectAll(selectAllCheckBox.getValue());
+    fireSelectionChangeEvent();
+  }
   /**
    * Gets the number of selected projects
    *
    * @return the number of selected projects
    */
   public int getSelectedProjectsCount() {
-    return selectedProjects.size();
+    if (folder != null) {
+      return folder.getSelectedProjects().size() + folder.getSelectedFolders().size();
+    }
+    else return 0;
   }
 
+
   public int getMyProjectsCount() {
     int count = 0;
-    for (Project project : projects) {
+    if (folder == null) {
+      return 0;
+    }
+    for (Project project : folder.getVisibleProjects()) {
       if (!project.isInTrash()) {
         ++ count;
       };
     }
     return count;
   }
 
-  /**
-   * Returns if the specified project is in Trash or in MyProjects
-   */
-  public int getProjectCurrentView(Project project) {
-    if (project.isInTrash()) {
-      return Ode.TRASHCAN;
+  public void setIsTrash(boolean isTrash) {
+    this.isTrash = isTrash;
+    if (isTrash) {
+      folder = Ode.getInstance().getFolderManager().getTrashFolder();
     } else {
-      return Ode.PROJECTS;
+      folder = Ode.getInstance().getFolderManager().getGlobalFolder();
+    }
+    if (folder != null) {
+      refresh();
     }
   }
 
-  /**
-   * Returns true if all projects under the current view have been selected, and false if not.
-   */
-  public boolean isAllProjectsSelected() {
-    if (Ode.getInstance().getCurrentView() == Ode.PROJECTS
-          && getSelectedProjectsCount() == getMyProjectsCount()) {
-      return true;
-    }
-    if (Ode.getInstance().getCurrentView() == Ode.TRASHCAN
-          && getSelectedProjectsCount() == projects.size() - getMyProjectsCount()) {
-      return true;
-    }
-    return false;
+  // FolderManagerEventListener implementation
+  @Override
+  public void onFolderAdded(ProjectFolder folder) {
+    refresh();
   }
-  /**
-   * Returns the list of selected projects.
-   *
-   * @return the selected projects
-   */
-  public List<Project> getSelectedProjects() {
-    return selectedProjects;
+
+  @Override
+  public void onFolderRemoved(ProjectFolder folder) {
+    refresh();
+  }
+
+  @Override
+  public void onFolderRenamed(ProjectFolder folder) {
+    refresh();
+  }
+
+  @Override
+  public void onFoldersChanged() {
+    refresh();
   }
 
-  //  ProjectManagerEventListener implementation
+  @Override
+  public void onFoldersLoaded() {
+    setIsTrash(isTrash);
+  }
 
   @Override
   public void onProjectAdded(Project project) {
-    projects.add(project);
-    projectWidgets.put(project, new ProjectWidgets(project));
-    if (!projectListLoading) {
-      refreshTable(true);
+    if (folder == null) {
+    }
+    if (projectsLoaded) {
+      folder.addProject(project);
+      Ode.getInstance().getFolderManager().saveAllFolders();
+      refresh();
```

Comment:
Note that `refresh()` translates into `refresh(false)` but in the original version it was `refreshTable(true)`, so the semantics have changed here. It might be the case that this doesn't end up sorting things correctly.

Reply:
> Note that `refresh()` translates into `refresh(false)` but in the original version it was `refreshTable(true)`, so the semantics have changed here. It might be the case that this doesn't end up sorting things correctly.

Huh. As far as I know, the sort functionality is working, but will verify. I think the needToSort flag functionality might have moved elsewhere.